tpu-provisioner/internal/controller/pods_started_controller.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/GoogleCloudPlatform/ai-on-gke/tpu-provisioner/internal/cloud"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// PodsStartedReconciler watches Pods and updates a ConfigMap when a minimum number of
+// associated Pods have started.
+// It assumes that these Pods are managed by a JobSet, otherwise it will ignore them.
+// The threshold is set via an annotation and is intended to be the total number of Pods
+// in the JobSet - allowing for semi-synchronous startup across an entire JobSet.
+// This threshold annotation is used to avoid a lookup of the number of replicas in the JobSet.
```

Comment:
If we only reconcile on leader pods (see comment below), I think it would be fine to look up the JobSet and derive the expected pod count from the spec, rather than hard coding it in the annotation. Maybe we can user a Lister to maintain a local cache with the JobSet object?

This would improve the user-experience IMO, we would need to validate it works at scale though.

Reply:
Completed